system/jlib/jmetrics.hpp

```diff
@@ -180,6 +192,31 @@ class jlib_decl GaugeMetric : public MetricVal
     }
 };
 
+
+class jlib_decl GaugeMetricFromCounters : public MetricVal
+{
+public:
+    GaugeMetricFromCounters(const char *name, const char *description, StatisticMeasure _units,
+                            const std::shared_ptr<CounterMetric> &_pBeginCounter, const std::shared_ptr<CounterMetric> &_pEndCounter,
+                            const MetricMetaData &_metaData = MetricMetaData()) :
+        MetricVal{name, description, MetricType::METRICS_GAUGE, _units, _metaData},
+        pBeginCounter{_pBeginCounter},
+        pEndCounter{_pEndCounter}
+    {
+        assert(pBeginCounter->queryUnits() == pEndCounter->queryUnits());
+    }
+
+    virtual __uint64 queryValue() const override
+    {
+        return pBeginCounter->queryValue() - pEndCounter->queryValue();
```

Comment:
really picky (I debated even mentioning it).
Better as

```
endValue = pEndCounter->queryValue()
return pBeginCounter->queryValue() - endValue;

```
Why?  The evaluation order is undefined in the previous code.  If begin->value was called, then both begin and end were incremented before end->value() was called you could end up with a -ve count.  Making the order explicit will ensure it is positive.

Reply:
An explicit order is better. The value can be 0 (which some argue is included with a positive designation).